lib/report_generators/search_term_by_dimensions.rb

```diff
@@ -0,0 +1,275 @@
+require 'date'
+require File.join(__dir__, '..', 'google_api_client')
+
+require File.join(File.dirname(__FILE__), '..', '..', 'config', 'app')
+
+class SearchTermByDimensions
+  attr_accessor :queries, :clicks, :dimensions
+
+  def initialize(options = {})
+    @google_api_client  = GoogleApiClient.new(auth_file: options[:auth_file])
+    @ga_profile_id = options[:ga_profile_id]
+    @start_date    = options[:start_date]
+    @end_date      = options[:end_date]
+    @output        = options[:output]
+    @google_parent_id = options[:google_parent_id]
+    @queries       = []
+    @clicks        = []
+  end
+
+  def report_basename
+    mapping = {
+      'today'     =>  Date.today.strftime(),
+      'yesterday' =>  (Date.today - 1).strftime()
+    }
+
+    real_start_date = mapping[@start_date] || @start_date
+    real_end_date   = mapping[@end_date] || @end_date
+
+    "output_#{real_start_date}_#{real_end_date}.csv"
+  end
+
+  def report_output_path
+    (@output == "google-sheets") ? File.join(File.absolute_path('.'), self.report_basename) : File.join(File.absolute_path(@output), self.report_basename)
+  end
+
+  def get_events!()
+    analytics_client = @google_api_client.analytics_client
+
+    query_response = analytics_client.get_ga_data(@ga_profile_id, @start_date, @end_date, 'ga:totalEvents,ga:uniqueEvents', dimensions: 'ga:eventLabel,ga:eventAction,ga:dimension1,ga:dimension2', max_results: 10000, filters: "ga:eventCategory==Search;ga:eventAction==QuerySent", sort: "-ga:totalEvents")
+    click_response = analytics_client.get_ga_data(@ga_profile_id, @start_date, @end_date, 'ga:totalEvents,ga:uniqueEvents,ga:avgEventValue', dimensions: 'ga:eventLabel,ga:eventAction,ga:dimension1,ga:dimension2', max_results: 10000, filters: "ga:eventCategory==Search;ga:eventAction==Clickthrough", sort: "ga:eventLabel")
+
+    if query_response.rows
+      query_response.rows.each do |query_row|
+        queries << QueryResponse.new({
+          search_term: query_row[0],
+          action: query_row[1],
+          searched_from: query_row[2],
+          searched_repo: query_row[3],
+          total_events: query_row[4].to_i,
+          unique_events: query_row[5].to_i
+        })
+      end
+    end
+
+    if click_response.rows
+      click_response.rows.each do |click_row|
+        clicks << ClickResponse.new({
+          search_term: click_row[0],
+          action: click_row[1],
+          searched_from: click_row[2],
+          searched_repo: click_row[3],
+          total_events: click_row[4].to_i,
+          unique_events: click_row[5].to_i,
+          mean_ordinality: click_row[6].to_f,
+        })
+      end
+    end
+  end
+
+  def process_data_for_term(term)
+    events_processor = TermEventsProcessor.new(term: term, query_segments: query_events_for_term(term), click_segments: click_events_for_term(term))
```

Comment:
I hand off per-term processing to a new class to encapsulate the logic of interrogating the segments.